server/lib/api/parties.ts

```diff
@@ -0,0 +1,182 @@
+
+const declineParamsSchema = partyIdSchema.keys({
+  targetId: Joi.number(),
+})
+
+export default function (router: Router) {
+  router
+    .use(ensureLoggedIn)
+    .post(
+      '/invites',
+      featureEnabled(PARTIES),
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ body: invitesPostSchema }),
+      invite,
+    )
+    .delete(
+      '/invites/:partyId/:targetId',
+      featureEnabled(PARTIES),
+      throttleMiddleware(invitesThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: declineParamsSchema }),
+      decline,
+    )
+    .post(
+      '/:partyId',
+      featureEnabled(PARTIES),
+      throttleMiddleware(partyThrottle, ctx => ctx.session!.userId),
+      joiValidator({ params: partyIdSchema }),
+      accept,
+    )
+}
+
+function isPartyServiceError(error: Error): error is PartyServiceError {
+  return error.hasOwnProperty('code')
+}
+
+function convertPartyServiceError(err: Error) {
+  if (!isPartyServiceError(err)) {
+    throw err
+  }
+
+  switch (err.code) {
+    case PartyServiceErrorCode.PartyNotFound:
+      throw new httpErrors.NotFound(err.message)
+    case PartyServiceErrorCode.InsufficientPermissions:
+      throw new httpErrors.Forbidden(err.message)
+    case PartyServiceErrorCode.PartyFull:
+      throw new httpErrors.Conflict(err.message)
+    case PartyServiceErrorCode.UserOffline:
+      throw new httpErrors.Unauthorized(err.message)
+    case PartyServiceErrorCode.InvalidAction:
+      throw new httpErrors.BadRequest(err.message)
+    default:
+      assertUnreachable(err.code as never)
+  }
+}
+
+async function invite(ctx: RouterContext) {
+  const { clientId, targets } = ctx.request.body
+
+  try {
+    const invites = await Promise.all<PartyUser>(
+      targets.map(
+        async (target: string): Promise<PartyUser> => {
+          const foundTarget = await users.find(target)
+          if (!foundTarget) {
+            throw new httpErrors.NotFound('Target user not found')
+          }
+
+          // TODO(2Pac): Check if the target user has blocked invitations from the user issuing
+          // the request. Or potentially use friends list when implemented.
+
+          return { id: foundTarget.id, name: foundTarget.name } as PartyUser
+        },
+      ),
+    )
+
+    const leader: PartyUser = {
+      id: ctx.session!.userId,
+      name: ctx.session!.userName,
+    }
+
+    partyService.invite(leader, clientId, invites)
+
+    ctx.status = 204
+  } catch (err) {
+    convertPartyServiceError(err)
+  }
+}
+
+async function decline(ctx: RouterContext) {
+  const { partyId, targetId } = ctx.params
+
+  try {
+    let leader: PartyUser | undefined
+    let target: PartyUser = { id: ctx.session!.userId, name: ctx.session!.userName }
+
+    // If the `targetId` is provided, that means the user issuing the request wants to cancel an
+    // invite to someone else.
+    if (targetId) {
+      const foundTarget = await users.find(targetId)
```

Comment:
`ctx.params` is an object of `string`s, so I don't this actually works as you'd expect. You need to parse the number out instead, otherwise `users.find` is going to search it as a username (which is in hindsight, a bad API :) )

Reply:
Hmm, I haven't tested this fully yet, but I thought `targetId` would be converted to a number by `Joi` since it was declared as a number in the schema.

Reply:
Nope, we don't anything with the output of Joi so no conversions occur.

I've been thinking that we should move to a model where we call a function to validate params inside the handlers (instead of doing it as part of a Koa middleware chain) but I haven't written the code for that yet. This would fix the need to do conversions by hand and also not be as hard to deal with in TypeScript (probably).